### DISCUSSION TITLE
Caption limited to 123 caracters ?

### DISCUSSION THREAD

**Discussion by fabienklesse** [ImageMagick/ImageMagick#3583] [Development]
Hi everyone, 
when I put a texte more than 123 caracters in a caption: , Imagemagick doesn't do anything. Do you have this issue ?
Best
Fabien

**Comment by fmw42** [comment]
What was your exact command line?  What is your Imagemagick version and your platform/OS?  Certain OS have limited number of characters in any string or command. If I recall, old versions of Windows were limited to 128 characters. But I willl defer to others to comment further or correct me.

**Reply by fabienklesse** [reply]
Hi, thanks for your answer,
i don't know the version but if you tell me that it's limited to 128 car, it's shurely that. I'm on a dedicated managed server on Ionos.

the line is this:


exec("/usr/bin/convert -font vavont.ttf -background white -size ".$largeurbackground."x".$hauteurbackground." -gravity northwest -extent +0-0 -pointsize 130 -fill '#111111' -density 43 -interline-spacing -10 caption:'Le jour de ses onze ans, Harry Potter, un orphelin élevé par un oncle et une tante qui le détestent? Et qui est l'effroyable V..., le mage dont personne n'ose prononcer le nom ?' +level 7%,100% -matte -virtual-pixel transparent -distort Perspective '0,0 1570,972 ".imagesx($dessouscoordonnees).",0 1160,930 ".imagesx($dessouscoordonnees).",".imagesy($dessouscoordonnees)." 1222,333 0,".imagesy($dessouscoordonnees)." 1600,379' ".$resize);

exec("/usr/bin/convert -quality 100 ".$dessous." -page +0+0 ".$resize." -flatten ".$final);

**Reply by fabienklesse** [reply]
but in my other scripts, i have more long command lines. 
is the command line limited to 128 car? or Imagemagick caption text ?

**Reply by fmw42** [reply]
Can you provide your command without actual values, not arguments?

**Reply by fmw42** [reply]
Can you remove parts of the command to find out where it fails?

**Reply by fabienklesse** [reply]
here is the line :

/usr/bin/convert -font vavont.ttf -background white -size 1920x1280 -gravity northwest -extent +0-0 -pointsize 130 -fill '#111111' -density 43 -interline-spacing -10 caption:'1234567890 1234567890 1234567890 1234567890 1234567890 1234567890' +level 7%,100% -matte -virtual-pixel transparent -distort Perspective '0,0 1570,972 1920,0 1160,930 1920,1280 1222,333 0,1280 1600,379' distort/resize466803068d30f66b84fc2bac1a07e8e2.png

if I add some 1234567890 , it fails. no error message, but it does not create the image.

**Comment by snibgo** [comment]
@fabienklesse: What version of IM, on what platform?

This command has over 200 characters in the caption:
```
magick caption:"abcdefghijklmnopqrstuvwxyz....abcdefghijklmnopqrstuvwxyz....abcdefghijklmnopqrstuvwxyz....abcdefghijklmnopqrstuvwxyz....abcdefghijklmnopqrstuvwxyz....abcdefghijklmnopqrstuvwxyz....abcdefghijklmnopqrstuvwxyz" x.png
```
![x](https://user-images.githubusercontent.com/33812368/115915197-ea572380-a46a-11eb-86d2-984fdc996d64.png)

Tested with M v7.0.8-64 on Windows 8.1.

**Reply by snibgo** [reply]
@fabienklesse: Does your command fail? What is the error message? Your caption contains the same quote character you use to delimit it. I expect that will cause a problem.

**Reply by fabienklesse** [reply]
no error message, image doesn't create if i put to many characters. 
```
/usr/bin/convert -font vavont.ttf -background white -size 1920x1280 -gravity northwest -extent +0-0 -pointsize 130 -fill '#111111' -density 43 -interline-spacing -10 caption:'1234567890 1234567890 1234567890 1234567890 1234567890 1234567890' +level 7%,100% -matte -virtual-pixel transparent -distort Perspective '0,0 1570,972 1920,0 1160,930 1920,1280 1222,333 0,1280 1600,379' distort/resize466803068d30f66b84fc2bac1a07e8e2.png
```

if I add some 1234567890 , it fails. no error message, but it does not create the image.

**Reply by fmw42** [reply]
Your command works fine for me in command line mode on IM 6.9.12.8 Q16 Mac OSX Sierra. I can even add 6 more sets of 1234567890 and it still works.  What is your IM version and platform/OS?

**Reply by fabienklesse** [reply]
i don't know the version, i use it on a dedicated managed server on ionos

**Reply by fmw42** [reply]
What do you get from:

`convert -version`

**Reply by fabienklesse** [reply]
absolutly nothing lol

**Reply by fabienklesse** [reply]
it works in command lines

**Reply by fmw42** [reply]
Sounds line some issue with Imagick not Imagemagick.

**Comment by snibgo** [comment]
@fabienklesse: Your exec() doesn't check for failure, or for error messages.

**Reply by fmw42** [reply]
As noted by snibgo, you can do the following:

```
<?php
exec("convert input commands output 2>&1",$out,$returnval);
foreach($out as $text)
{echo "$text<br>";}
?>
```

To get error messages from exec calling Imagemagick. Put you actual Imagemagick convert command in place of "convert input commands output"

**Reply by fabienklesse** [reply]
wow thanks here is the message :

convert-im6.q16: unable to open image `input': No such file or directory @ error/blob.c/OpenBlob/2874.
convert-im6.q16: no decode delegate for this image format `' @ error/constitute.c/ReadImage/560.
convert-im6.q16: unable to open image `commands': No such file or directory @ error/blob.c/OpenBlob/2874.
convert-im6.q16: unable to open image `commands': No such file or directory @ error/blob.c/OpenBlob/2874.
convert-im6.q16: no decode delegate for this image format `' @ error/constitute.c/ReadImage/560.
convert-im6.q16: no images defined `output' @ error/convert.c/ConvertImageCommand/3258.

**Reply by fmw42** [reply]
The input is the caption that you create. So that message does not make sense to me. Perhaps your /tmp directory is full from all the 'crashes' and needs to be emptied.

**Reply by snibgo** [reply]
@fabienklesse: Use your actual ImageMagick command, not the example "convert input commands output" that fmw42 used.

**Reply by fabienklesse** [reply]
sorry for my ignorance , i try with it

exec("/usr/bin/convert -font vavont.ttf -background white -size ".$largeurbackground."x".$hauteurbackground." -gravity northwest -extent +0-0 -pointsize 130 -fill '#111111' -density 43 -interline-spacing -10 caption:'1234567890 1234567890 1234567890 1234567890 1234567890 1234567890' +level 7%,100% -matte -virtual-pixel transparent -distort Perspective '0,0 1570,972 ".imagesx($dessouscoordonnees).",0 1160,930 ".imagesx($dessouscoordonnees).",".imagesy($dessouscoordonnees)." 1222,333 0,".imagesy($dessouscoordonnees)." 1600,379' ".$resize." 2>&1",$out,$returnval);
foreach($out as $text)
{echo "$text<br>";}

it writes nothing

**Reply by fabienklesse** [reply]
ahhh sorry, 
here is the message when I go up to 123 car

convert: width or height exceeds limit `1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890 1234567890' @ error/caption.c/ReadCAPTIONImage/161.
convert: no images defined `distort/resize466803068d30f66b84fc2bac1a07e8e2.png' @ error/convert.c/ConvertImageCommand/3258.

**Reply by fmw42** [reply]
What do you get from 

```
<?php
exec("convert -list resource 2>&1",$out,$returnval);
foreach($out as $text)
{echo "$text<br>";}
?>

```
and

```
<?php
exec("convert -version 2>&1",$out,$returnval);
foreach($out as $text)
{echo "$text<br>";}
?>
```

**Reply by fmw42** [reply]
Your Imagemagick default policy.xml file likely is limited to 128 characters and is uncommented.

 
You will need to ask your hosting provide to increase it or comment it as below:

`
`

**Reply by fabienklesse** [reply]
Resource limits:
Width: 16KP
Height: 16KP
List length: 18.446744EP
Area: 128MP
Memory: 256MiB
Map: 512MiB
Disk: 1GiB
File: 6144
Thread: 8
Throttle: 0
Time: unlimited



Version: ImageMagick 6.9.10-23 Q16 x86_64 20190101 https://imagemagick.org
Copyright: © 1999-2019 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP
Delegates (built-in): bzlib djvu fftw fontconfig freetype heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png tiff webp wmf x xml zlib

**Reply by fabienklesse** [reply]
arf ok 128 max.... thanks a lot fmw42 ;)

**Reply by fmw42** [reply]
That is just my suspicion. But your resources say List length: 18.446744EP. So I may be wrong. What is EP?  E pixels? must be extremely big?

Ask your hosting provider to see if it needs increasing.  Do a test first with exactly 127 characters including spaces and then 128 and then 129.  Does it break at 128 or 129?

**Reply by fabienklesse** [reply]
Hey Snibgo is right !!!

**Comment by snibgo** [comment]
@fabienklesse: You have "-pointsize 130". Try with a smaller size, eg 10.

**Reply by fabienklesse** [reply]
Thank you so much !!! i've set pointsize at 10 and density at 500 and it works !!!!!!!!
Thanks a lot Snibgo

**Reply by fmw42** [reply]
@snibgo, please explain why that works. Is that trading file size for image height? Was his text exceeding the height of the image allowed by his resources?

**Reply by fabienklesse** [reply]
the text was stopped after 123 or 128 caracters and now it works perfectly 

Before, the text was stopped
![1396010813_1619199229](https://user-images.githubusercontent.com/83067149/115936350-4e8edd00-a495-11eb-908d-eae7495e6398.jpg)

and now
![1396010813_1619217188](https://user-images.githubusercontent.com/83067149/115936416-74b47d00-a495-11eb-9fe1-c01eed333a58.jpg)

**Comment by snibgo** [comment]
With 123 characters at pointsize 130, the error message was:
```
convert: width or height exceeds limit 
```
The caption has many characters, so the limit exceeded was likely to be the width, not the height.
```
magick -pointsize 130 caption:M info:

caption:M CAPTION 109x147 109x147+0+0 16-bit sRGB 0.094u 0:00.069
```
So one character is 109 pixels wide, and 123 characters is about 13407, which is suspiciously close to the delegates.xml limit of 16 KB.

**Reply by fmw42** [reply]
But caption should wrap. So why is the width too large?

**Reply by snibgo** [reply]
In caption.c function ReadCAPTIONImage(), checks that pointsize multiplied by the number of characters in the string does not exceed the width resource limit.

130*123 = 16974, which is greater than the limit of 16K, so a "WidthOrHeightExceedsLimit" exception is thrown.

The check may be more conservative than it needs to be. Perhaps that check could be removed, and replaced by more sophisticated checks later in the code.

**Reply by fmw42** [reply]
I agree that it is too conservative. An approximate wrap should be considered. 

On the other hand, the OP has specified no width or height for his caption image. Perhaps he should specify a desired width or desired height in addition to his pointsize. Would that help avoid this issue?

**Reply by snibgo** [reply]
The check is made whether or not "-size" is specified.

We can show that the test is too conservative by setting a width limit of 200, and a caption of 10 characters:
```
f:\web\im>%IMG7%magick -limit width 200 -pointsize 19 caption:1234567890 info:
caption:1234567890 CAPTION 111x24 111x24+0+0 16-bit sRGB 0.172u 0:00.172

f:\web\im>%IMG7%magick -limit width 200 -pointsize 20 caption:1234567890 info:
magick: width or height exceeds limit `1234567890' @ error/caption.c/ReadCAPTIONImage/163.
```
Increasing the pointsize from 19 to 20 should increase the required width from 111 to about 111*20/19 = 117, which is less than 200, so the test is too strict.

However, "caption:" is complex and calls AnnotateImage(), and replacing that check with a minimum check, while ensuring the width limit is never exceeded, would not be simple (I think).

**Comment by urban-warrior** [comment]
We introduced sanity checks to guard against nefarious command-lines that hang indefinitely.  We'll have a patch for the issue within a day or two.  Thanks for alerting us to this issue.
